### ISSUE TITLE
Recovering from apparent server errors when using itis_terms with many names

### ISSUE THREAD

**Issue by chuckrp** [ropensci/taxize#271]
taxize is terrific, very useful. However, when I use itis_terms with multiple names (300+) I frequently have problems. Three things tend to happen which I am unable to recover from. The most common is an error message that the server return was empty. That's not too bad as I can resubmit until it finally runs. A very rare problem is a time out when the server does not respond appropriately. A serious and not uncommon problem is that the process hangs. When this happens my only recourse is to force quit R, which is never a good option.

I assume all of these are server side errors. However, it would be nice if taxize could catch at least the server returned empty error and elegantly resolve or bypass the problem. Also, a few days ago there was a post about another taxize function that had a problem with large numbers of names, and a solution was presented. I am unable to relocate that issue, but perhaps that solution would help clear up this problem with itis_terms.

Chuck

**Comment by chuckrp** [comment]
Scott,
Thank you for handling this issue. There has been a long hiatus since I first brought it up, but I am now retired and have had time to get back to what I was trying to do back then. First of all, this is just an update for your interest. I do not want to reopen #217.

My goal was to check the status of, get TSNs for, and find authors and common names where available for 40,328 names from a biodiversity database I am involved with. Because of the number of names, I split the list into 50 submissions of 800 and 1 submission of 328 names. When submitted to ITIS, I received the "server returns empty" error at a rate of about every third or fourth 800 name submission. Much better than I expected, actually. Resubmitting from each stop point allowed me to eventually get all the names processed. I followed the same procedure to submitthe names to GBIF and EOL. GBIF responded without any problems at all. EOL also responded well, for the most part. However, EOL did choke on a few names, stopping when they were submitted. It seemed to be an issue with non-standard characters within some of the names, probably a result of original data-entry errors. Removing those particular entries resulted in good returns. GBIF apparently just
 ignored those names and returned a not found result.

As far as the returns are concerned, each group of 800 names returned as many as 1,700,000 records. I was astounded. However, after removing hundreds of thousands of duplicates each group of 800 names resulted in 2,000-5,000 unique names. ITIS provided common names for about 5,000 species. On the other hand, when compared with my original list, ITIS matched 24,328, EOL matched 34,041, and GBIF matched 35,420. This does not account for valid/accepted/synonym and so on.

I would like to make a request regarding sci2comm. Namely, to be able to run it with ask=FALSE and have it return all found common names for each taxon. Sitting at the monitor while thousands of names are processed is not a viable option for much of what I do. Plus, there is nothing wrong with multiple common names. I noted that ITIS allows for at least as many as 39. If you wish, I can submit this through GitHub to get it into the system.

Taxize is a great tool, based on these results. I realize it may not be the 
type of use you had in mind when you developed it, but taxize has been 
extremely helpful to me.
Chuck




On Friday, November 28, 2014 9:55 AM, Scott Chamberlain <notifications@github.com> wrote:
 


Closed #271.

**Comment by sckott** [comment]
hi @chuckrp  Congrats on retiring!

Sorry about the troubles with ITIS.  Are you interested in trying local querying of ITIS data?  I am working on a development version of taxize that works with local copies of databases, so far ITIS, NCBI, COL, and Theplantlist. If so, let me know and I'll help you get that set up. 

For `sci2comm` I think what you want is the new functions with the underscore on the end. e.g., 

```r
get_tsn_('Poa an')
```

gives

```r
$`Poa an`
     tsn    scientificname      commonnames nameusage
2  41107         Poa annua annual bluegrass  accepted
5 784054 Nicoraepoa andina             <NA>  accepted
6 784648          Poa anae             <NA>  accepted
7 784649     Poa androgyna             <NA>  accepted
```

Without any prompts. The normal function `get_tsn()` would give a prompt with this query, but this function does not, just gives all data back.  

Unfortuantely, you can't pass that directly to `sci2comm`, but you could coerce each tsn into a proper id object like

```r
res <- get_tsn_('Poa an')
sci2comm(as.tsn(res$`Poa an`$tsn))
```

gives 

```r
$`41107`
[1] "annual blue grass" "walkgrass"         "annual bluegrass" 

$`784054`
[1] NA

$`784648`
[1] NA

$`784649`
[1] NA
```

thx for the kind works, glad it's been useful!
